eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/OfferCodecs.scala

```diff
@@ -46,11 +47,31 @@ object OfferCodecs {
     b => if (b) 0 else 1
   )
 
-  private val shortChannelIdDirCodec: Codec[ShortChannelIdDir] =
+  private val shortChannelIdDirCodec: Codec[EncodedNodeId.ShortChannelIdDir] =
     (("isNode1" | isNode1) ::
-      ("scid" | realshortchannelid)).as[ShortChannelIdDir]
+      ("scid" | realshortchannelid)).as[EncodedNodeId.ShortChannelIdDir]
+
+  private val plainNodeIdCodec: Codec[EncodedNodeId.WithPublicKey.Plain] = Codec[EncodedNodeId.WithPublicKey.Plain](
+    (nodeId: EncodedNodeId.WithPublicKey.Plain) => bytes(33).encode(nodeId.publicKey.value),
+    (wire: BitVector) => bytes(33).decode(wire).flatMap(res => res.value.head match {
+      case 0x02 | 0x03 => Attempt.successful(res.map(bin => EncodedNodeId.WithPublicKey.Plain(PublicKey(bin))))
+      case d => Attempt.Failure(new Err.MatchingDiscriminatorNotFound(d))
+    })
+  )
+
+  private val walletNodeIdCodec: Codec[EncodedNodeId.WithPublicKey.Wallet] = Codec[EncodedNodeId.WithPublicKey.Wallet](
+    (nodeId: EncodedNodeId.WithPublicKey.Wallet) => bytes(33).encode((nodeId.publicKey.value.head + 2).toByte +: nodeId.publicKey.value.tail),
+    (wire: BitVector) => bytes(33).decode(wire).flatMap(res => res.value.head match {
+      case 0x04 | 0x05 => Attempt.successful(res.map(bin => EncodedNodeId.WithPublicKey.Wallet(PublicKey((bin.head - 2).toByte +: bin.tail))))
```

Comment:
This is the non-standard part, the risk of incompatibility seems very low.

Reply:
AFAIK there is currently no plan to use this in a conflicting way. If someone does propose something, we'll just ask them to use a different discriminator than `0x04` and `0x05` :crossed_fingers: 

By the way I couldn't find a better way of writing those codecs, but I'm not very satisfied...

Reply:
> we'll just ask them to use a different discriminator than 0x04 and 0x05

The discriminator is *before* the public key right? It would be a different encoding for public key.

Reply:
No, that's the issue, the discriminator is the first byte of the public key...this was initially a way to extend the codec for public keys:

- if the first byte is `0x00` or `0x01`, this cannot be a public key, and is instead a directed scid
- if the first byte is `0x02` or `0x03`, this is a public key, and should be treated as such
- (what we're adding) if the first byte is `0x04` or `0x05`, this is a public key for a mobile wallet node *if you replace 0x04 by 0x02 and 0x05 by 0x03*

Reply:
> By the way I couldn't find a better way of writing those codecs, but I'm not very satisfied...

Challenge accepted 😆 

```scala
  /** A 32-bytes codec for public key where the first byte is set manually. */
  def tweakFirstByteCodec(prefix: Byte): Codec[PublicKey] = bytes(32).xmap(b => PublicKey(prefix +: b), _.value.drop(1))

  val encodedNodeIdCodec: Codec[EncodedNodeId] = discriminated[EncodedNodeId].by(uint8)
    .subcaseP(0x00) { case e: EncodedNodeId.ShortChannelIdDir if e.isNode1 => e }(realshortchannelid.xmap(EncodedNodeId.ShortChannelIdDir(true, _), _.scid))
    .subcaseP(0x01) { case e: EncodedNodeId.ShortChannelIdDir if !e.isNode1 => e }(realshortchannelid.xmap(EncodedNodeId.ShortChannelIdDir(false, _), _.scid))
    .subcaseP(0x02) { case e: EncodedNodeId.WithPublicKey.Plain if e.publicKey.value.head == 0x02 => e }(tweakFirstByteCodec(2).xmap[EncodedNodeId.WithPublicKey.Plain](EncodedNodeId.WithPublicKey.Plain, _.publicKey))
    .subcaseP(0x03) { case e: EncodedNodeId.WithPublicKey.Plain if e.publicKey.value.head == 0x03 => e }(tweakFirstByteCodec(3).xmap[EncodedNodeId.WithPublicKey.Plain](EncodedNodeId.WithPublicKey.Plain, _.publicKey))
    .subcaseP(0x04) { case e: EncodedNodeId.WithPublicKey.Wallet if e.publicKey.value.head == 0x02 => e }(tweakFirstByteCodec(2).xmap[EncodedNodeId.WithPublicKey.Wallet](EncodedNodeId.WithPublicKey.Wallet, _.publicKey))
    .subcaseP(0x05) { case e: EncodedNodeId.WithPublicKey.Wallet if e.publicKey.value.head == 0x03 => e }(tweakFirstByteCodec(3).xmap[EncodedNodeId.WithPublicKey.Wallet](EncodedNodeId.WithPublicKey.Wallet, _.publicKey))
```

Reply:
![witchcraft](https://github.com/ACINQ/eclair/assets/31281497/63b49f7b-050b-4dfb-9192-9b61b8b0cd9b)

Reply:
Nice one, it took me a while to understand the first partial function argument (why it's using `0x02` and `0x03` in the `Wallet` case), but now it makes sense and is much better than the `choice` codec we were previously using!